File: src/Application.Resources/scripts/event/q3239.lua
local entryMap;
local exitMap;
local eventLength = 20;

function init()
    em:setProperty("noEntry", "false");
    entryMap = em:GetMap(922000000);
    exitMap = em:GetMap(922000009);
end

function setup(level, lobbyid)
    local eim = em:newInstance("q3239_" + lobbyid);
    eim:setExclusiveItems({ 4031092 });
    return eim;
end

function playerEntry(eim, player)
    local im = eim:getInstanceMap(entryMap:getId());

    -- Reset instance
    im:clearDrops();
    im:resetReactors();
    im:shuffleReactors();

    -- Start timer
    eim:startEventTimer(eventLength * 60 * 1000);

    -- Warp player and mark event as occupied
    player:changeMap(entryMap, 0);
    em:setProperty("noEntry", "true");
end

function changedMap(eim, player, mapid)
    if (mapid ~= entryMap:getId()) then
        playerExit(eim, player);
    end
end

function playerExit(eim, player)
    finish(eim);
end

function playerDisconnected(eim, player)
    finish(eim);
end

function scheduledTimeout(eim)
    finish(eim);
end

function finish(eim)
    local party = eim:getPlayers(); -- should only ever be one player
    for _, player in ipairs(party) do
        eim:unregisterPlayer(player);
        player:changeMap(exitMap);
    end

    eim:dispose();
    em:setProperty("noEntry", "false");
end

-- Stub/filler functions

function disbandParty(eim, player)
end

function afterSetup(eim) 
end

function playerUnregistered(eim, player)
end

function changedLeader(eim, leader) 
end

function leftParty(eim, player) 
end

function clearPQ(eim) 
end

function dispose() 
end

function cancelSchedule() 
end

function allMonstersDead(eim) 
end

function monsterValue(eim, mobId) 
end

function monsterKilled(mob, eim) 
end

File: src/Application.Resources/scripts/event/q3239.js
var name = "q3239";
var eventType = "Solo";
var entryMap = 922000000;
var exitMap = 922000009;
var eventLength = 20;
const maxLobbies = 7;

function init() {
}

function getMaxLobbies() {
    return maxLobbies;
}

function setup(level, lobbyid) {
    var eim = em.newInstance("q3239_" + lobbyid);
    eim.setExclusiveItems([4031092]);
    return eim;
}

function playerEntry(eim, player) {
    var im = eim.getInstanceMap(entryMap);

    // Reset instance
    im.clearDrops();
    im.resetReactors();
    im.shuffleReactors();

    // Start timer
    eim.startEventTimer(eventLength * 60 * 1000);

    // Warp player and mark event as occupied
    player.changeMap(im, 0);
}

function changedMap(eim, player, mapid) {
    if (mapid != entryMap)
        playerExit(eim, player);
}

function playerExit(eim, player) {
    end(eim);
}

function playerDisconnected(eim, player) {
    end(eim);
}

function scheduledTimeout(eim) {
    end(eim);
}

function end(eim) {
    var party = eim.getPlayers(); // should only ever be one player
    for (var i = 0; i < party.Count; i++) {
        var player = party[i];
        eim.unregisterPlayer(player);
        player.changeMap(exitMap);
    }

    eim.dispose();
}

// Stub/filler functions

function disbandParty(eim, player) {}
function afterSetup(eim) {}
function playerUnregistered(eim, player) {}
function changedLeader(eim, leader) {}
function leftParty(eim, player) {}
function clearPQ(eim) {}
function dispose() {}
function cancelSchedule() {}
function allMonstersDead(eim) {}
function monsterValue(eim, mobId) {}
function monsterKilled(mob, eim) {}
